### PR TITLE
Project controller is reconciling on gcm update

### PR DESCRIPTION
**Pull request by swilen-iwanow**
**What this PR does / why we need it**: 
Project controller is reconciling projects on gardener controller manager restart. 

**Which issue(s) this PR fixes**:
Fixes: Project resources were not reconciled on chart update

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Gardener now reconciles project resources on restart.
```

### CODE REVIEW

File: pkg/apis/garden/types.go [hunk 1]
```diff
@@ -492,6 +492,8 @@ type ProjectSpec struct {
 
 // ProjectStatus holds the most recently observed status of the project.
 type ProjectStatus struct {
+	// Gardener holds information about the Gardener which last acted on the Shoot.
```
Comment (vpnachev) [inline pkg/apis/garden/types.go]:
```suggestion
	// Gardener holds information about the Gardener which last acted on the Project.
```

File: pkg/apis/garden/v1beta1/types.go [hunk 2]
```diff
@@ -489,6 +489,9 @@ type ProjectSpec struct {
 
 // ProjectStatus holds the most recently observed status of the project.
 type ProjectStatus struct {
+	// Gardener holds information about the Gardener which last acted on the Shoot.
```
Comment (vpnachev) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
	// Gardener holds information about the Gardener which last acted on the Project.
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
A copy-paste typo.

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks for the PR.

In general, it definitely makes sense to reconcile the `Project` resource when GCM starts. I have some doubts about the implementation via the status field, though. If we had the `project.Generation != project.Status.ObservedGeneration` in the [projectUpdate](https://github.com/gardener/gardener/blob/44840998ebb7ee7e89f00dc94525d8fa3a163386/pkg/controllermanager/controller/project/project_control.go#L46) we wouldn't have to add a new status field.

**Comment by vpnachev** [comment]
A few months back this PR #690 was merged, but on our landscapes there were projects that were not reconciled which means that the clusterrole `garden.sapcloud.io:system:project-member:<project-name>` was not updated. All of this happened because the `project.Generation == project.Status.ObservedGeneration` hence, the project controller did not reconciled the Project resources. @adracus is aware of this issue and he manually updated the projects specs to force the reconciliation.

So my opinion is that we cannot keep using only the generation and observedGeneration to determine whether e reconciliation is needed or not.

File: pkg/controllermanager/controller/project/project_control.go [hunk 3]
```diff
@@ -125,9 +127,11 @@ func (c *defaultControl) ReconcileProject(obj *gardenv1beta1.Project) (bool, err
 	if project.DeletionTimestamp != nil {
 		return c.delete(project, projectLogger)
 	}
-	if project.Generation != project.Status.ObservedGeneration {
+
+	if !apiequality.Semantic.DeepEqual(project.Status.Gardener, c.identity) || project.Generation != project.Status.ObservedGeneration {
```
Comment (vpnachev) [inline pkg/controllermanager/controller/project/project_control.go]: I would separate the conditions 
```suggestion
	if !apiequality.Semantic.DeepEqual(project.Status.Gardener, c.identity) {
		return false, c.reconcile(project, projectLogger)
	}
	
	if project.Generation != project.Status.ObservedGeneration {
		return false, c.reconcile(project, projectLogger)
	}
	
```

**Comment by timuthy** [comment]
> So my opinion is that we cannot keep using only the generation and observedGeneration to determine 
> whether e reconciliation is needed or not.

So your change makes Gardener reconcile the `Project` resource every time GCM starts (and only then because the fields of Gardener struct won't change at runtime) which might be a desired behaviour and I'm totally fine with that.

My proposal is to rather use the `projectAdd` event handler to accomplish that. It is called by the SharedInformer for every `Project` resource on start-up. Therefore, you'd need to move `project.Generation != project.Status.ObservedGeneration` to `projectUpdate` because `ReconcileProject` will probably also be called from `projectAdd`.

With that in place we don't need no enhancement of the status resource.

**Comment by vpnachev** [comment]
Thanks @timuthy ! Now I get your proposal and it definitely makes sense.
@swilen-iwanow could you adapt your PR to what Tim propose?

File: pkg/controllermanager/controller/project/project_control.go [hunk 4]
```diff
@@ -44,6 +45,15 @@ func (c *Controller) projectAdd(obj interface{}) {
 }
 
 func (c *Controller) projectUpdate(oldObj, newObj interface{}) {
+	var (
+		oldProject = oldObj.(*gardenv1beta1.Project)
```
Comment (vpnachev) [inline pkg/controllermanager/controller/project/project_control.go]:
```suggestion
```

File: pkg/controllermanager/controller/project/project_control.go [hunk 5]
```diff
@@ -24,6 +24,7 @@ import (
 	kutils "github.com/gardener/gardener/pkg/utils/kubernetes"
 
 	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
```
Comment (vpnachev) [inline pkg/controllermanager/controller/project/project_control.go]:
```suggestion

```

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: pkg/controllermanager/controller/project/project_control.go [hunk 6]
```diff
@@ -44,6 +45,15 @@ func (c *Controller) projectAdd(obj interface{}) {
 }
 
 func (c *Controller) projectUpdate(oldObj, newObj interface{}) {
+	var (
+		oldProject = oldObj.(*gardenv1beta1.Project)
+		newProject = newObj.(*gardenv1beta1.Project)
+	)
+
+	if apiequality.Semantic.DeepEqual(oldProject.Generation, newProject.Generation) {
```
Comment (vpnachev) [inline pkg/controllermanager/controller/project/project_control.go]:
```suggestion
	if newProject.Generation == newProject.Status.ObservedGeneration {
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
I forgot one more thing that could be changed.

File: pkg/controllermanager/controller/project/project_control.go (hunk 6 above)
Comment (swilen-iwanow) [reply pkg/controllermanager/controller/project/project_control.go]: Actually we can check for equality on the newProject vs oldProject objects on update. That way we can get rid of the observedGeneration in the status, if it doesn't introduce any issues (API incompatibility) on existing resources that are being fetch from etcd. @timuthy what are your thoughts about this?

Comment (plkokanov) [reply pkg/controllermanager/controller/project/project_control.go]: Generally I think that comparing against `ObservedGeneration` is better. But lets wait for @timuthy.

`ObservedGeneration` is incremented by the gardener controller manager only if reconciliation is successful. 

`Generation` gets incremented by the gardener apiserver when an update request is processed for the corresponding resource and only if there is a difference between the specs of the previous and new resources.

Comment (timuthy) [reply pkg/controllermanager/controller/project/project_control.go]: Technically, we can also call `Semantic.DeepEqual(oldProject, newProject)` but I think we don't benefit from that.

The `Generation` is already increased if  `Semantic.DeepEqual(oldProject, newProject) == false` (see https://github.com/gardener/gardener/blob/3501ea171f3963edcf88f7afa7c73aa6c7ea89ac/pkg/registry/garden/project/strategy.go#L61) and already belongs to the API which is why we also shouldn't remove it now (compatibility). I think it's also considered a Kubernetes best practice to have an `ObservedGeneration` when the resource has a `status` sub-resource.

Bottom line, `newProject.Generation == newProject.Status.ObservedGeneration` is probably cheaper and simpler in this case.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by timuthy — APPROVED** [review]
/lgtm
